Make exec_transfer a valid option for an up-to-date execution before sending data to sub apps

GiudGiud opened the pull request:
as the regolds show, we were lagging data accidentally in the user objects transfers.

refs #28981

I am hitting all the UO/PP transfers. Note that an auxkernel executed on transfers (not an option at this time) would make sense for transfers sending from aux variables

GiudGiud committed:
Add execution of UOs on transfer for UO field transfers

Add exec_transfer to the PP transfers
closes #28981

Add option for spatial UOs to execute on transfers
Dont forget PRE_AUX in PP and UO transfer exec

We dont support transfer PRE_IC, and no UO I know is useful PRE_IC to create something to transfer

GiudGiud committed:
Add exec_transfer to a few transfers

GiudGiud committed:
Ban exec_transfer on the multiapp side
- it's a tad confusing because it s a parent schedule in a subapp
- it's not necessary, subapp should already be updated if simply using timestep_end

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2929236/), step Clang format on ae078ac wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30694/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30694/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format eca7ccbcecd170c2a7a35edc16f8532bbcd1937e`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2929321/), step Docs: sync website on 81fc62d wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30694/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2929336/), step Generate coverage on 81fc62d wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>eca7cc</th>
      <th colspan=4>#30694 81fc62</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.53%</td>
      <td>85.53%</td>
      <td>+0.00%</td>
      <td>93.22%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>113556</td>
      <td>113609</td>
      <td>+53</td>
      <td>55</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>19218</td>
      <td>19224</td>
      <td>+6</td>
      <td>4</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30694/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30694/coverage/framework)


## Modules coverage

### Thermal hydraulics

<table>
  <thead>
    <tr>
      <th></th>
      <th>eca7cc</th>
      <th colspan=4>#30694 81fc62</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.81%</td>
      <td>88.81%</td>
      <td>+0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>14898</td>
      <td>14899</td>
      <td>+1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1878</td>
      <td>1878</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30694/coverage_diff/thermal_hydraulics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30694/coverage/thermal_hydraulics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/30694/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented:
we probably should consider a warning on missing ExEC_TRANSFER in the PPs/UOs. I might attempt this when addressing the review. I imagine it would require fixing hundreds of tests though

lindsayad commented on framework/include/userobjects/SpatialUserObjectFunctor.h:
```diff
@@ -20,7 +20,14 @@ template <typename UserObjectType>
 {
 public:
-  static InputParameters validParams() { return UserObjectType::validParams(); }
+  static InputParameters validParams()
+  {
+    auto params = UserObjectType::validParams();
```
move outside the class definition

lindsayad commented on framework/include/transfers/MultiAppTransfer.h:
```diff
@@ -213,6 +213,17 @@ class MultiAppTransfer : public Transfer
   }
 
+  /**
+   * Error if executing this MooseObject in the source multiapp on EXEC_TRANSFER. We error because:
+   * - it should be enough to execute this object on TIMESTEP_END to have an updated value
+   * - EXEC_TRANSFER in the child app is behavior that is not identical when ran independently
```
Also it seems like this method is only ever called with `is_from_multiapp = true`. So should we just remove that parameter?

lindsayad replied:
Based off looking at the rest of the PR, especially if we remove the boolean parameter, it seems like we can just say
> Error if executing this MooseObject on EXEC_TRANSFER in a source from multiapp

GiudGiud replied:
ok I ll simplify this a bit. 
I implemented the is_from_multiapp = false in every transfer which is why I dont need that case

lindsayad commented on framework/src/transfers/MultiAppTransfer.C:
```diff
@@ -656,3 +656,33 @@ MultiAppTransfer::getLocalSourceAppIndex(unsigned int i_from) const
+      .query()
+      .template condition<AttribName>(object_name)
+      .queryInto(objects_with_exec_on);
+  for (const auto obj_ptr : objects_with_exec_on)
+    if (obj_ptr->getExecuteOnEnum().contains(EXEC_TRANSFER))
```
I would use `AttribExecOns` to avoid doing this manually

lindsayad replied:
Unless you really want that later assertion

GiudGiud committed:
Address Alex's review
- simplify error on exec_transfer in wrong app check
- simplify search in warehouse

lindsayad commented on framework/include/transfers/MultiAppTransfer.h:
```diff
@@ -213,6 +213,14 @@ class MultiAppTransfer : public Transfer
   }
 
+  /**
+   * Error if executing this MooseObject on EXEC_TRANSFER in a source application (from_multiapp)
```
```suggestion
   * Error if executing this MooseObject on EXEC_TRANSFER in a source multiapp (from_multiapp, e.g. child/sibling app)
```

lindsayad commented on framework/include/transfers/MultiAppTransfer.h:
```diff
@@ -213,6 +213,14 @@ class MultiAppTransfer : public Transfer
 
+  /**
+   * Error if executing this MooseObject on EXEC_TRANSFER in a source application (from_multiapp)
+   * Note that when the parent app is the source application, it is usually desired to use
```
```suggestion
   * Note that, conversely, when the parent app is the source application, it is usually \emph desired to use
```

GiudGiud committed:
Improve docstrings

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2956131/), step Clang format on 1118d17 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30694/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30694/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 6bf6ef7148917edd6a24002615085afccde71b10`

moosebuild commented:
Job [Public apps](https://civet.inl.gov/job/2961109/) on 717c008 : invalidated by @lindsayad

moosebuild commented:
Job [Clang](https://civet.inl.gov/job/2961110/) on 717c008 : invalidated by @lindsayad
